ofnetSvcProxy.go

```diff
@@ -0,0 +1,662 @@
+/***
+Copyright 2016 Cisco Systems Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ofnet
+
+import (
+	"errors"
+	"strconv"
+	"net"
+	"github.com/shaleman/libOpenflow/protocol"
+	"github.com/shaleman/libOpenflow/openflow13"
+	"sync"
+	"github.com/contiv/ofnet/pqueue"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/contiv/ofnet/ofctrl"
+)
+
+// service proxy implementation
+
+const (
+	watchedFlowMax = 2
+	spDNAT = "Dst"
+	spSNAT = "Src"
+)
+
+// PortSpec defines protocol/port info required to host the service
+type PortSpec struct {
+	Protocol string
+	SvcPort uint16  // advertised port
+	ProvPort uint16  // actual port of provider
+}
+
+// ServiceSpec defines a service to be proxied
+type ServiceSpec struct {
+	IpAddress string
+	Ports []PortSpec
+}
+
+// Providers holds the current providers of a given service
+type Providers struct {
+	providers map[string]bool  // Provider IP as key
+}
+
+// svcCatalogue holds information about all services to be proxied
+// Accessible by north-bound API
+type svcCatalogue struct {
+	svcMap map[string]ServiceSpec  // service name as key
+	provMap map[string]Providers   // service name as key
+}
+
+// provOper holds operational info for each provider
+type provOper struct {
+	clientEPs map[string]bool	// IP's of endpoints served by the provider
+	pqHdl *pqueue.Item	// handle into the providers pq
+}
+
+// proxyOper is operational state of the proxy
+type proxyOper struct {
+	ports []PortSpec
+	provHdl map[string]provOper  // provider IP as key
+	provPQ *pqueue.MinPQueue  // provider priority queue for load balancing
+	watchedFlows []*ofctrl.Flow  // flows this service is watching
+	natFlows map[string]*ofctrl.Flow // epIP.[in|out] as key
+}
```

Comment:
just for my education, how are these proxy states restored after a netplugin restart?